File: libmeasure/measure.c
//
//  General-purpose Photovoltaic Device Model gpvdm.com- a drift diffusion
//  base/Shockley-Read-Hall model for 1st, 2nd and 3rd generation solarcells.
// 
//  Copyright (C) 2012-2017 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
//
//	https://www.gpvdm.com
//	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
//
//
// This program is free software; you can redistribute it and/or modify it
// under the terms and conditions of the GNU General Public License,
// version 2, as published by the Free Software Foundation.
//
// This program is distributed in the hope it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.


#include "util.h"
#include "const.h"
#include "light.h"
#include "device.h"
#include "const.h"
#include <dump.h>
#include "config.h"
#include "inp.h"
#include "util.h"
#include "cal_path.h"
#include "lang.h"
#include "log.h"
#include "measure.h"
#include <probe.h>
#include <buffer.h>
#include <sys/stat.h>
#include <i.h>
#include <rpn.h>

static int unused __attribute__((unused));

struct probe_config config;

static struct istruct spectrum_first;
static int first=FALSE;

void measure_file(struct simulation *sim,char *file_name)
{
	int i;
	int vector;
	int enable;
	char vec_str[400];
	char file[400];
	char data_path[400];
	long double position;
	char output_token[400];
	char temp[400];
	char sim_name[400];

	char math[400];
	FILE *out;
	struct inp_file inp;
	struct inp_file inp_data;
	char input_position[400];
	char output_file[400];

	struct rpn rpn_cal;
	long double value;
	long double y;
	int write=FALSE;
	int ret=0;
	strcpy(vec_str,"");

	join_path(2, file,get_input_path(sim),file_name);

	inp_init(sim,&inp);
	if (inp_load(sim, &inp , file)!=0)
	{
		ewe(sim,"I can't find file %s\n",file);
	}

	inp_check(sim,&inp,1.0);
	inp_reset_read(sim,&inp);
	struct istruct data;
	inter_init(sim,&data);

	inp_get_string(sim,&inp);	//name
	strcpy(sim_name,inp_get_string(sim,&inp));	//

	inp_get_string(sim,&inp);	//enable
	strcpy(temp,inp_get_string(sim,&inp));
	enable=english_to_bin(sim,temp);

	if (enable==FALSE)
	{
		inp_free(sim,&inp);
		return;
	}

	inp_get_string(sim,&inp);	//compile to vector
	strcpy(temp,inp_get_string(sim,&inp));
	vector=english_to_bin(sim,temp);

	sprintf(output_file,"measure_%s.dat",sim_name);

	out=fopena(get_output_path(sim),output_file,"w");


	while(1)
	{
		write=FALSE;

		strcpy(temp,inp_get_string(sim,&inp));	//file token
		if (strcmp(temp,"#ver")==0)
		{
			break;
		}

		strcpy(file_name,inp_get_string(sim,&inp));		//file
		
		inp_get_string(sim,&inp);	//position token
		strcpy(input_position,inp_get_string(sim,&inp));
		
		inp_get_string(sim,&inp);	//output
		
		strcpy(output_token,inp_get_string(sim,&inp));	//output token
		//printf("%s %Lf %s\n",file_name,position,output_token);

		join_path(2,data_path,get_output_path(sim),file_name);

		inp_get_string(sim,&inp);
		strcpy(math,inp_get_string(sim,&inp));	//output token
		
		if (strcmp_begin(input_position,"#")==0)
		{
				inp_init(sim,&inp_data);
				if (inp_load(sim, &inp_data , data_path)==0)
				{
					ret=inp_search_gdouble(sim,&inp_data,&value,input_position);

					if (ret==0)
					{
						rpn_init(sim,&rpn_cal);
						rpn_add_var(sim,&rpn_cal,"x",(double)value);

						y=(long double)rpn_evaluate(sim,&rpn_cal,math);
						//printf("%Le %s %Le %s\n",y,math,value,input_position);
						//getchar();
						inp_free(sim,&inp_data);
						write=TRUE;
					}else
					{
						printf_log(sim,_("Token not found: %s"),input_position);
					}

				}else
				{
					printf_log(sim,_("File not found: %s"),data_path);
				}
		}else
		{
			if (inter_load(sim,&data,data_path)==0)
			{
				inter_sort(&data);
				//inter_dump(sim,&data);
				sscanf(input_position,"%Le",&(position));
				y=inter_get_hard(&data,position);

				rpn_init(sim,&rpn_cal);
				rpn_add_var(sim,&rpn_cal,"x",(double)y);

				y=(long double)rpn_evaluate(sim,&rpn_cal,math);
				
				inter_free(&data);
				write=TRUE;
			}
		}
		
		if (write==TRUE)
		{
			if (vector==FALSE)
			{
				fprintf(out,"#%s\n",output_token);
				fprintf(out,"%Le\n",y);
			}else
			{
				sprintf(temp,"%Le ",y);
				strcat(vec_str,temp);
			}
		}

	}

	inp_free(sim,&inp);

	if (vector==FALSE)
	{
		fprintf(out,"#ver\n");
		fprintf(out,"#1.0\n");
		fprintf(out,"#end\n");
	}else
	{
		fprintf(out,"%s",vec_str);	
	}

	fclose(out);
}


void measure(struct simulation *sim)
{
int i=0;
struct inp_list a;
inp_listdir(sim,get_input_path(sim),&a);


	for (i=0;i<a.len;i++)
	{
		if ((strcmp(a.names[i],".")!=0)&&(strcmp(a.names[i],"..")!=0))
		{
			if ((cmpstr_min(a.names[i],"measure")==0)&&(strcmp_end(a.names[i],".inp")==0))
			{
				measure_file(sim,a.names[i]);
			}
		}
	}

inp_list_free(&a);
}

File: gui/measure.py
#    General-purpose Photovoltaic Device Model - a drift diffusion base/Shockley-Read-Hall
#    model for 1st, 2nd and 3rd generation solar cells.
#    Copyright (C) 2012-2017 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
#
#	https://www.gpvdm.com
#	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License v2.0, as published by
#    the Free Software Foundation.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along
#    with this program; if not, write to the Free Software Foundation, Inc.,
#    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


import os

from gui_util import dlg_get_text
from scan_select import select_param
from token_lib import tokens
from scan_item import scan_items_get_list

from plot_io import plot_save_oplot_file
from scan_io import scan_push_to_hpc
from scan_io import scan_import_from_hpc
from cal_path import get_exe_command
from icon_lib import QIcon_load
from scan_item import scan_items_get_file
from scan_item import scan_items_get_token
from util import str2bool

from scan_item import scan_items_lookup_item
from gui_util import tab_move_down
from gui_util import tab_move_up

import i18n
_ = i18n.language.gettext

#qt
from PyQt5.QtCore import QSize, Qt 
from PyQt5.QtWidgets import QWidget,QVBoxLayout,QToolBar,QSizePolicy,QAction,QTabWidget,QTableWidget,QAbstractItemView, QMenuBar,QTableWidgetItem
from PyQt5.QtGui import QPainter,QIcon

from gui_util import tab_add
from gui_util import tab_remove
from gui_util import tab_get_value

from inp import inp_save_lines_to_file
from inp import inp_load_file
from inp import inp_isfile
from inp import inp_copy_file
from inp import inp_update_token_value
from inp import inp_get_token_value
from inp import inp_remove_file

from gpvdm_select import gpvdm_select

from scan_select import select_param

from cal_path import get_sim_path
from QWidgetSavePos import QWidgetSavePos
from window_list import resize_window_to_be_sane
from measure_ribbon import measure_ribbon
from measure_tab import measure_tab
from QHTabBar import QHTabBar
from util_zip import zip_lsdir

from util import strextract_interger

from gui_util import yes_no_dlg

def measure_new_filename():
	for i in range(0,20):
		pulse_name="measure"+str(i)+".inp"
		if inp_isfile(os.path.join(get_sim_path(),pulse_name))==False:
			return i
	return -1

class measure(QWidgetSavePos):


	def load_tabs(self):

		file_list=zip_lsdir(os.path.join(get_sim_path(),"sim.gpvdm"))
		files=[]
		for i in range(0,len(file_list)):
			if file_list[i].startswith("measure") and file_list[i].endswith(".inp"):
				name=inp_get_token_value(file_list[i], "#measurement_name")
				files.append([name,file_list[i]])

		files.sort()

		for i in range(0,len(files)):
			value=strextract_interger(files[i][1])
			if value!=-1:
				self.add_page(value)

	def add_page(self,index):
		tab=measure_tab(index)
		name=inp_get_token_value(tab.file_name, "#measure_name")
		self.notebook.addTab(tab,name)

	def __init__(self):
		QWidgetSavePos.__init__(self,"measure_window")
		resize_window_to_be_sane(self,0.5,0.7)

		self.setWindowIcon(QIcon_load("measure"))
		self.setWindowTitle(_("Measurment editor")+" (https://www.gpvdm.com)")


		self.vbox=QVBoxLayout()

		self.ribbon=measure_ribbon()
		self.ribbon.tb_new.triggered.connect(self.callback_add_page)
		self.ribbon.tb_rename.triggered.connect(self.callback_rename_page)
		self.ribbon.tb_clone.triggered.connect(self.callback_copy_page)
		self.ribbon.tb_delete.triggered.connect(self.callback_delete_page)		


		self.vbox.addWidget(self.ribbon)

		self.notebook = QTabWidget()
		self.notebook.setTabBar(QHTabBar())
		self.notebook.setTabPosition(QTabWidget.West)
		self.notebook.setMovable(True)

		self.vbox.addWidget(self.notebook)

		self.load_tabs()

		self.setLayout(self.vbox)


	def callback_add_page(self):
		new_sim_name=dlg_get_text( _("New measurement name")+":", _("measurement ")+str(self.notebook.count()+1),"document-new.png")

		if new_sim_name.ret!=None:
			index=measure_new_filename()
			inp_copy_file(os.path.join(get_sim_path(),"measure"+str(index)+".inp"),os.path.join(get_sim_path(),"measure0.inp"))
			inp_update_token_value(os.path.join(get_sim_path(),"measure"+str(index)+".inp"), "#measure_name", new_sim_name.ret)
			self.add_page(index)

	def callback_copy_page(self):
		tab = self.notebook.currentWidget()

		old_index=tab.index
		new_sim_name=dlg_get_text( _("Clone the current measurement to a new measurement called:"), _("measurement ")+str(self.notebook.count()+1),"clone.png")
		new_sim_name=new_sim_name.ret
		if new_sim_name!=None:
			index=measure_new_filename()
			if inp_copy_file(os.path.join(get_sim_path(),"measure"+str(index)+".inp"),os.path.join(get_sim_path(),"measure"+str(old_index)+".inp"))==False:
				print(_("Error copying file")+"measure"+str(old_index)+".inp")
				return

			inp_update_token_value(os.path.join(get_sim_path(),"measure"+str(index)+".inp"), "#measure_name", new_sim_name)
			self.add_page(index)

	def remove_invalid(self,input_name):
		return input_name.replace (" ", "_")

	def callback_rename_page(self):
		tab = self.notebook.currentWidget()
		name=inp_get_token_value(tab.file_name, "#measure_name")

		new_sim_name=dlg_get_text( _("Rename the measurement to be called")+":", name,"rename.png")

		new_sim_name=new_sim_name.ret

		if new_sim_name!=None:
			tab.rename(new_sim_name)
			index=self.notebook.currentIndex() 
			self.notebook.setTabText(index, new_sim_name)


	def callback_delete_page(self):

		tab = self.notebook.currentWidget()
		name=inp_get_token_value(tab.file_name, "#measure_name")

		response=yes_no_dlg(self,_("Should I remove the measurment file ")+name)

		if response == True:
			inp_remove_file(os.path.join(get_sim_path(),tab.file_name))
			index=self.notebook.currentIndex() 
			self.notebook.removeTab(index)
